### PR TITLE
Adds React.Fragment support

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

This PR adds React.Fragment support and updates `react`/`react-dom` Jest test dependencies.

### CODE REVIEW

File: src/serializer/ResidualReactElementSerializer.js [hunk 1]
```diff
@@ -252,7 +252,24 @@ export class ResidualReactElementSerializer {
...
       this.residualHeapSerializer.serializeValue(reactLibraryObject);
     }
-    let identifier = convertExpressionToJSXIdentifier(this.residualHeapSerializer.serializeValue(typeValue), true);
+    let identifier;
+    if (typeValue instanceof SymbolValue) {
+      // if there is no React library, then we should throw and error, as it is needed for React.Fragment output
+      if (reactLibraryObject === undefined) {
+        throw new FatalError("unable to serialize JSX fragment due to React not being referenced in scope");
+      }
+      // we want to vist the Symbol type, but we don't want to serialize it
```
Comment (gaearon) [inline src/serializer/ResidualReactElementSerializer.js]: There are more things described as symbols. We shouldn't just assume that if it's a symbol, it must be a fragment. We should throw for unknown ones.

Comment (trueadm) [reply src/serializer/ResidualReactElementSerializer.js]: Address in a new commit

**Comment by gaearon** [comment]
Can we add tests for state preservation when switching between Fragments and arrays? See tests added as part of merging Fragment export.

**Comment by trueadm** [comment]
@gaearon If you have any tests in mind, feel free to paste them on this PR and I'll add more tests around similar flavours.

File: test/react/functional-components/key-not-change-fragments.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+App.getTrials = function(renderer, Root) {
+  let results = [];
+  renderer.update(<Root switch={false} />);
+  results.push(['mount', renderer.toJSON()]);
+
+  renderer.update(<Root switch={false} />);
+  results.push(['update with same key', renderer.toJSON()]);
+
+  renderer.update(<Root switch={true} />);
+  results.push(['update with different key', renderer.toJSON()]);
```
Comment (gaearon) [inline test/react/functional-components/key-not-change-fragments.js]: These descriptions don't match what we're testing anymore. They should say "update with array" or "update with fragment".

**Comment by gaearon** [comment]
Here is a test that still produces `Symbol.for` in the output. Is that expected?

```js
var Stateful = (function (superclass) {
  function Stateful () {
    superclass.apply(this, arguments);
    this.state = { updated: false };
  }

  if ( superclass ) {
    Stateful.__proto__ = superclass;
  }
  Stateful.prototype = Object.create( superclass && superclass.prototype );
  Stateful.prototype.constructor = Stateful;
  Stateful.prototype.componentWillReceiveProps = function componentWillReceiveProps() {
    this.setState({ updated: true });
  }
  Stateful.prototype.render = function render () {
    return (
      <div>
        {this.props.children}
        (is update: {String(this.state.updated)})
      </div>
    );
  };

  return Stateful;
}(React.Component));

function Stateless(props) {
  return <Stateful x={props.x}>{props.children}</Stateful>;
}

function App(props) {
  if (props.switch) {
    return (
      <React.Fragment>
        <Stateless key='hi' x={props.x}>Hi</Stateless>
      </React.Fragment>
    );
  }
  return [
    <Stateless key='hi' x={props.x}>Bye</Stateless>
  ];
}
```

**Comment by trueadm** [comment]
@gaearon I don't see any mention of `Symbol.for` in my compiled output? :/

**Comment by gaearon** [comment]
Here is a full test:

```js


var React = require('react');
// the JSX transform converts to React, so we need to add it back in
this['React'] = React;

var Stateful = (function (superclass) {
  function Stateful () {
    superclass.apply(this, arguments);
    this.state = { updated: false };
  }

  if ( superclass ) {
    Stateful.__proto__ = superclass;
  }
  Stateful.prototype = Object.create( superclass && superclass.prototype );
  Stateful.prototype.constructor = Stateful;
  Stateful.prototype.componentWillReceiveProps = function componentWillReceiveProps() {
    this.setState({ updated: true });
  }
  Stateful.prototype.render = function render () {
    return (
      <div>
        {this.props.children}
        (is update: {String(this.state.updated)})
      </div>
    );
  };

  return Stateful;
}(React.Component));

function Stateless(props) {
  return <Stateful x={props.x}>{props.children}</Stateful>;
}

function App(props) {
  if (props.switch) {
    return (
      <React.Fragment>
        <Stateless key='hi' x={props.x}>Hi</Stateless>
      </React.Fragment>
    );
  }
  return [
    <Stateless key='hi' x={props.x}>Bye</Stateless>
  ];
}


App.getTrials = function(renderer, Root) {
  let results = [];
  renderer.update(<Root switch={false} />);
  results.push(['mount', renderer.toJSON()]);

  renderer.update(<Root switch={false} />);
  results.push(['update with same type', renderer.toJSON()]);

  renderer.update(<Root switch={true} />);
  results.push(['update with different type', renderer.toJSON()]);

  renderer.update(<Root switch={true} />);
  results.push(['update with same type (again)', renderer.toJSON()]);

  renderer.update(<Root switch={false} />);
  results.push(['update with different type (again)', renderer.toJSON()]);
  return results;
};

if (this.__registerReactComponentRoot) {
  __registerReactComponentRoot(App);
}

module.exports = App;
```

Here is a part of `compiledCode` that looks odd to me:

```js
      var _$g = _$c.Symbol;
      var _$h = _$g.for;

      // ...

        var _7 = _D(_$h("react.fragment"), null, _c);
```

File: src/serializer/ResidualReactElementSerializer.js [hunk 3]
```diff
@@ -252,7 +273,15 @@ export class ResidualReactElementSerializer {
...
     }
-    let identifier = convertExpressionToJSXIdentifier(this.residualHeapSerializer.serializeValue(typeValue), true);
+    let identifier;
+    if (typeValue instanceof SymbolValue && typeValue === getReactSymbol("react.fragment", this.realm)) {
+      identifier = convertExpressionToJSXIdentifier(
+        this._serializeReactFragmentType(typeValue, reactLibraryObject),
+        true
+      );
+    } else {
+      identifier = convertExpressionToJSXIdentifier(this.residualHeapSerializer.serializeValue(typeValue), true);
```
Comment (gaearon) [inline src/serializer/ResidualReactElementSerializer.js]: Nit: this code would be clearer if we kept the function call unconditional, and only branched for the first argument

File: test/react/functional-components/key-not-change-fragments.js (hunk 2 above)
Comment (gaearon) [reply test/react/functional-components/key-not-change-fragments.js]: Still an issue

**Comment by gaearon** [comment]
Please fix the test description before landing: https://github.com/facebook/prepack/pull/1413#discussion_r167058851

**Pull request closed without merging**